Packs/CortexXDR/Integrations/XQLQueryingEngine/XQLQueryingEngine.py

```diff
@@ -0,0 +1,767 @@
+    query = args.get('query', '')
+    if not query:
+        raise ValueError('query is not specified')
+    if '//' in query:
+        raise DemistoException('Please remove notes (//) from query')
+
+    if 'limit' not in query:  # if user did not provide a limit in the query, we will use the default one.
+        query = f'{query} | limit {str(DEFAULT_LIMIT)}'
+    data = {
+        'request_data': {
+            'query': query,
+        }
+    }
+    time_frame = args.get('time_frame')
+    if time_frame:
+        converted_time = convert_relative_time_to_milliseconds(time_frame)
+        data['request_data']['timeframe'] = {'relativeTime': converted_time}
+    tenant_ids = argToList(args.get('tenant_ids'))
+    if tenant_ids:
+        data['request_data']['tenants'] = tenant_ids
+    # Call the Client function and get the raw response
+    execution_id = client.start_xql_query(data)
+    return execution_id
+
+
+def get_xql_query_results(client: Client, args: dict) -> Tuple[dict, Optional[bytes]]:
+    """Retrieve results of an executed XQL query API. returns the general response and
+    a file data if the query has more than 1000 results.
+
+    :type client: ``Client``
+    :param client: The XDR Client.
+    :type args: ``dict``
+    :param args: The arguments to pass to the API call.
+
+    :return: The query results.
+    :rtype: ``dict`` or tuple
+    """
+
+    query_id = args.get('query_id')
+    if not query_id:
+        raise ValueError('query ID is not specified')
+    format_method = args.get('format', 'json')
+    limit = args.get('limit', DEFAULT_LIMIT)
+    data = {
+        'request_data': {
+            'query_id': query_id,
+            'pending_flag': True,
+            'format': format_method,
+            'limit': limit
+        }
+    }
+
+    # Call the Client function and get the raw response
+    response = client.get_xql_query_results(data)
+    response['execution_id'] = query_id
+    results = response.get('results', {})
+    stream_id = results.get('stream_id')
+    if stream_id:
+        file_data = get_query_result_stream(client, stream_id)
+        return response, file_data
+    response['results'] = results.get('data')
+    return response, None
+
+
+def get_query_result_stream(client: Client, stream_id: str) -> bytes:
+    """Retrieve XQL query results with more than 1000 results.
+
+    :type client: ``Client``
+    :param client: The XDR Client.
+    :type stream_id: ``str``
+    :param stream_id: The stream ID of the query.
+
+    :return: The query results.
+    :rtype: ``bytes``
+    """
+
+    if not stream_id:
+        raise ValueError('stream_id is not specified')
+    data = {
+        'request_data': {
+            'stream_id': stream_id,
+            'is_gzip_compressed': True,
+        }
+    }
+    # Call the Client function and get the raw response
+    return client.get_query_result_stream(data)
+
+
+def format_results(list_to_format: list, remove_empty_fields: bool = True) -> list:
```

Comment:
This is a pretty messy function.
Think of a different implementation for it, and look at CommonServerPython to see if we have any relevant functions there.